src/Angular.js

```diff
@@ -901,13 +909,13 @@ function equals(o1, o2) {
         if (!isArray(o2)) return false;
         if ((length = o1.length) == o2.length) {
           for (key = 0; key < length; key++) {
-            if (!equals(o1[key], o2[key])) return false;
+            if (!equals(o1[key], o2[key], cache)) return false;
           }
           return true;
         }
       } else if (isDate(o1)) {
         if (!isDate(o2)) return false;
-        return equals(o1.getTime(), o2.getTime());
+        return equals(o1.getTime(), o2.getTime(), cache);
```

Comment:
Unrelated to this PR, but since `getTime()` returns a number, couldn't we just return `o1.getTime() === o2.getTime()` ?